optuna/integration/botorch.py

```diff
@@ -0,0 +1,468 @@
+            [
+                "torch.Tensor",
+                "torch.Tensor",
+                Optional["torch.Tensor"],
+                "torch.Tensor",
+            ],
+            "torch.Tensor",
+        ] = None,
+        constraints_func: Optional[
+            Callable[
+                [
+                    "MultiObjectiveStudy",
+                    "FrozenMultiObjectiveTrial",
+                ],
+                Sequence[float],
+            ]
+        ] = None,
+        n_startup_trials: int = 10,
+        independent_sampler: Optional[BaseMultiObjectiveSampler] = None,
+    ):
+        _imports.check()
+
+        self._candidates_func = candidates_func
+        self._constraints_func = constraints_func
+        self._independent_sampler = independent_sampler or RandomMultiObjectiveSampler()
+        self._n_startup_trials = n_startup_trials
+
+        self._trial_constraints: Dict[int, Tuple[float, ...]] = {}
+        self._study_id: Optional[int] = None
+        self._search_space = IntersectionSearchSpace()
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        if self._study_id is None:
+            self._study_id = study._study_id
+        if self._study_id != study._study_id:
+            # Note that the check below is meaningless when `InMemortyStorage` is used
+            # because `InMemortyStorage.create_new_study` always returns the same study ID.
+            raise RuntimeError("BoTorchSampler cannot handle multiple studies.")
+
+        return self._search_space.calculate(study, ordered_dict=True)  # type: ignore
+
+    def _update_trial_constraints(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trials: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> None:
+        # Since trial constraints are computed on each worker, constraints should be computed
+        # deterministically.
+
+        assert self._constraints_func is not None
+
+        for trial in trials:
+            number = trial.number
+            if number not in self._trial_constraints:
+                constraints = self._constraints_func(study, trial)
+
+                if not isinstance(constraints, (tuple, list)):
+                    raise TypeError("Constraints must be a tuple or list.")
+
+                constraints = tuple(constraints)
+                self._trial_constraints[number] = constraints
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        assert isinstance(search_space, OrderedDict)
+
+        trials = [t for t in study.get_trials(deepcopy=False) if t.state == TrialState.COMPLETE]
```

Comment:
Good point. How about doing that in a follow up? I think we should accompany that with some experimental results.

Reply:
I am not super familiar with the optuna internals, but there are two things that might be worth considering.
1. If you have multiple concurrent processes and optuna is asked to generate a new HP config, it's important that you let BoTorch know that you have "pending points" that are awaiting computation.
2. Adding partial evaluations to the dataset without telling BoTorch that the data is from a partial evaluation will cause problems with the model.  If you wanted to include partial evaluation, you'd want to include another input dimension that tells you, for example, the % of the training that was completed before the results were returned, and probably use a different acquisition function, like [KG](https://github.com/facebook/Ax/blob/master/ax/models/torch/botorch_kg.py) or MES.

Reply:
@eytan Thanks for your insightful comments.

1. Optuna will only tell BoTorch about trials that have already been completed, so BoTorch will have less information available than if it were completely sequential, but we think this is a compromise we should accept. Other samplers make the same compromise.
2. The idea of adding another dimension to the objective function to represent the progress of the optimization seems very reasonable to me. This is one of the so-called multi-fidelity optimization approaches, and I think it would be a similar device to the https://arxiv.org/abs/1605.07079. (It's "contextual optimization" as it's called in bandit problems and Bayesian optimization.)

In any case, I agree with the opinion that consideration of these problems should be a follow-up.

Reply:
1. I see, so this is a technical limitation of Optuna?  I'd be concerned that if this information about pending points weren't passed into BoTorch any type of parallel async evaluation will end up sampling the same set of points over and over again.
2. Yes, exactly.  The Knowledge Gradient implementation I linked to follows the [Practical Multi-fidelity Bayesian Optimization for Hyperparameter Tuning](http://proceedings.mlr.press/v115/wu20a/wu20a.pdf) paper, and we also have an [implementation](https://github.com/facebook/Ax/blob/9c546f7d5f15bb50b9eda4837a52a1d4f22d8f67/ax/models/torch/botorch_modular/mes.py) of [Multi-fidelity Max-value Entropy Search](http://proceedings.mlr.press/v119/takeno20a/takeno20a.pdf).  In BoTorch, we refer to the context variables as "fixed features" as they are not variables being optimized over.  We also support passing in multiple "trace observations" (samples from the learning curve), although this functionality hasn't been very thoroughly tested.

Reply:
> ... any type of parallel async evaluation will end up sampling the same set of points over and over again

This is very useful information. Actually, we should be able to construct a tensor of "pending points" (since parameters that have been suggested but not evaluated are persisted and available to all workers) so this might be something that we can look into. In my understanding, we basically have to pass this information to the acquisition function. That'd probably require a change to the interface of the callback to the sampler but it should be acceptable given that this is currently marked experimental.

Reply:
1. Thank you for the useful information. This is not a technical limitation of Optuna, but rather a limitation of the current PR implementation of BoTorchSampler: Optuna can access parameters that have already been sampled and are being evaluated. If BoTorch can distinguish between parameters that are being evaluated and those that have already been evaluated, then we can work around this problem by devising a better implementation of BoTorchSampler. However, for this PR, I think it would be smart to handle this as a follow-up, without making that effort.
2. That's interesting information. The hyperparameter optimization framework Optuna might be a good practical example of multi-fidelity optimization! I'm looking forward to BoTorchSampler supporting multi-fidelity optimization in the near future as a follow-up task.